File: show-off-api/python/src/services/queue_service.py
import logging

from werkzeug.datastructures import FileStorage

from models.detection_api import MinimalTextBox
from models.image_api import ImageBoxData, TextCorrection
from queues.queues import ImageQueue, ImageDataQueue, TextCorrectionQueue


log = logging.getLogger(__name__)


class QueueService:

    def __init__(self):
        self.image_queue = ImageQueue()
        self.image_data_queue = ImageDataQueue()
        self.text_correction_queue = TextCorrectionQueue()

    def send_image(self, request_id: str, file: FileStorage):
        log.info(f'Sending request with id {request_id} to queue {self.image_queue.queue}')
        self.image_queue.send(QueueService.__read_bytes_from_file(file), QueueService.__create_request_header(request_id))

    def send_image_data(self, request_id: str, boxes: list[MinimalTextBox], texts: list[str]):
        if len(boxes) != len(texts):
            log.warning(f'For request {request_id} there is missmatch in length of boxes and texts')
            return
        request = list(map(lambda box, text: ImageBoxData(box.start_x, box.start_y, box.width, box.height, text.text),
                           boxes, texts))
        log.info(f'Sending request with id {request_id} to queue {self.image_data_queue.queue}')
        self.image_data_queue.send(request, QueueService.__create_request_header(request_id))

    def send_text_correction(self, request_id: str, text_correction: str):
        log.info(f'Sending request with id {request_id} to queue {self.text_correction_queue.queue}')
        self.text_correction_queue.send(TextCorrection(request_id, text_correction),
                                        QueueService.__create_request_header(request_id))

    @staticmethod
    def __read_bytes_from_file(file: FileStorage):
        file.stream.seek(0)
        file_bytes = file.read()
        file.stream.seek(0)
        return file_bytes

    @staticmethod
    def __create_request_header(request_id: str):
        return {'request_id': request_id}

File: show-off-api/golang/main/service/queue_service.go
package service

import (
	"encoding/json"
	"github.com/streadway/amqp"
	"golang/main/api"
	"golang/main/common"
	"log"
	"os"
)

type QueueService struct {
	Channel *amqp.Channel
	Queues  map[string]QueueConnectionData
}

func NewQueueService(server ServerConnectionData, queues []QueueConnectionData) *QueueService {
	service := new(QueueService)
	connectionString := "amqp://" + server.Username + ":" + server.Password + "@" + server.Host + ":" + server.Port + "/" + server.VirtualHost
	connection, err := amqp.Dial(connectionString)
	if err != nil {
		panic(err)
	}
	channel, err := connection.Channel()
	if err != nil {
		panic(err)
	}
	queueMap := make(map[string]QueueConnectionData, 0)
	for _, queueData := range queues {
		_, err := channel.QueueDeclare(
			queueData.QueueName,
			true,
			false,
			false,
			false,
			nil,
		)
		if err != nil {
			panic(err)
		}
		queueMap[queueData.QueueName] = queueData
	}
	service.Channel = channel
	service.Queues = queueMap
	return service
}

func (service QueueService) SendImage(requestId string, image []byte) {
	queue := service.Queues[os.Getenv("IMAGE_QUEUE_NAME")]
	err := service.sendDataToQueue(queue, requestId, image, "application/octet-stream")
	failOnError(err, "Did not send data to imageQueue")
}

func (service QueueService) SendImageData(requestId string, boxes api.MinimalTextBoxes, tokens []string) {
	data := ImageBoxDataList{}
	if len(tokens) != len(boxes) {
		return
	}
	for i := range boxes {
		data = append(data, ImageBoxData{
			StartX: boxes[i].StartX,
			StartY: boxes[i].StartY,
			Width:  boxes[i].Width,
			Height: boxes[i].Height,
			Text:   tokens[i],
		})
	}
	queue := service.Queues[os.Getenv("IMAGE_DATA_QUEUE_NAME")]
	err := service.sendDataToQueue(queue, requestId, data, "application/json")
	failOnError(err, "Did not send data to imageDataQueue")
}

func (service QueueService) SendTextCorrection(inputRequest common.TextCorrectionRequest) {
	request := TextCorrection{Id: inputRequest.Id, Text: inputRequest.Text}
	queue := service.Queues[os.Getenv("TEXT_CORRECTION_QUEUE_NAME")]
	err := service.sendDataToQueue(queue, request.Id, request, "application/json")
	failOnError(err, "Did not send data to TextCorrectionQueue")
}

func (service QueueService) sendDataToQueue(queue QueueConnectionData, requestId string, payload interface{}, contentType string) error {
	byteRequest, err := json.Marshal(payload)
	if err != nil {
		return err
	}
	return service.Channel.Publish(
		queue.Exchange,
		queue.QueueName,
		false,
		false,
		amqp.Publishing{
			ContentType: contentType,
			Headers:     map[string]interface{}{"request_id": requestId},
			Body:        byteRequest,
		})
}

func failOnError(err error, msg string) {
	if err != nil {
		log.Fatalf("%s: %s", msg, err)
	}
}
